engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```
@@ -361,59 +608,225 @@ private String getProperty(long snapshotId, String property) {
         CopyCommandResult result = new CopyCommandResult(null, null);
+
         result.setResult("Unsupported operation requested for copying data.");
+
         callback.complete(result);
     }
+
+    private Map<String, String> getDetails(DataObject dataObj) {
+        if (dataObj instanceof VolumeInfo) {
+            return getVolumeDetails((VolumeInfo)dataObj);
+        }
+        else if (dataObj instanceof SnapshotInfo) {
+            return getSnapshotDetails((SnapshotInfo)dataObj);
+        }
+
+        throw new CloudRuntimeException("'dataObj' must be of type 'VolumeInfo' or 'SnapshotInfo'.");
+    }
+
+    private CopyCmdAnswer performResignature(DataObject dataObj, HostVO hostVO) {
+        return performResignature(dataObj, hostVO, false);
+    }
+
+    private CopyCmdAnswer performResignature(DataObject dataObj, HostVO hostVO, boolean keepGrantedAccess) {
+        long storagePoolId = dataObj.getDataStore().getId();
+        DataStore dataStore = _dataStoreMgr.getDataStore(storagePoolId, DataStoreRole.Primary);
+
+        Map<String, String> details = getDetails(dataObj);
+
+        ResignatureCommand command = new ResignatureCommand(details);
+
+        ResignatureAnswer answer = null;
+
+        try {
+            _volumeService.grantAccess(dataObj, hostVO, dataStore);
+
+            answer = (ResignatureAnswer)_agentMgr.send(hostVO.getId(), command);
+        }
+        catch (Exception ex) {
+            keepGrantedAccess = false;
+
+            throw new CloudRuntimeException(ex.getMessage());
+        }
+        finally {
+            if (keepGrantedAccess == false) {
+                try {
+                    _volumeService.revokeAccess(dataObj, hostVO, dataStore);
+                }
+                catch (Exception ex) {
+                    s_logger.debug(ex.getMessage(), ex);
```

1. Why not log this ignored exception to ``WARN``?  It seems like the operator would want to know if we failed to rollback the operation.
1. Please add some contextual information around the exception message to explain that we failed to rollback access to the datastore on the host.